### PR TITLE
Add -on-error command line argument to allow preserving artifacts on builder errors

### LINKED ISSUE #409: --no-destroy-on-error like Vagrant
It would appear that an error exit code by `postinstall.sh` is enough to totally wipe out the generated boxes.

It would be useful to keep them around to manually manipulate while working on them.  The `-debug` switch can be used for this, but it's not really ideal since you basically have to know the appropriate step (`stepCreateVM`) to wait at.

See also: https://github.com/mitchellh/vagrant/issues/2011

### PR DESCRIPTION
**Pull request by orivej**
This pull request resolves #409 by providing an option to delay or skip cleanup after unsuccessful `packer build`:

```
  -on-error=cleanup          When a builder fails, clean up and exit (the default)
  -on-error=abort            When a builder fails, abort without cleanup
  -on-error=ask              When a builder fails, prompt for action
```

Running with `-on-error=ask`:

```
==> qemu: Step "StepProvision" failed
==> qemu: [C]lean up and exit, [A]bort without cleanup, or [R]etry step (build may fail even if retry succeeds)? [car] |
```

### CODE REVIEW

**Comment by rickard-von-essen** [comment]
Looks interesting! This is missing some documentation. (Currently the build is busted, you can ignore Travis)

File: builder/azure/arm/builder.go [hunk 1]
```diff
@@ -157,7 +157,7 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
 		ui.Message(fmt.Sprintf("temp admin password: '%s'", b.config.Password))
 	}
 
-	b.runner = b.createRunner(&steps, ui)
+	b.runner = packerCommon.NewRunner(steps, b.config.PackerConfig, ui)
```
Comment (boumenot) [inline builder/azure/arm/builder.go]: I really like how this code is simplified for every builder.

File: common/multistep_runner.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	"reflect"
+	"time"
+
+	"github.com/mitchellh/multistep"
+	"github.com/mitchellh/packer/packer"
+)
+
+func newRunner(steps []multistep.Step, config PackerConfig, ui packer.Ui) (multistep.Runner, multistep.DebugPauseFn) {
+	switch config.PackerOnError {
+	case "cleanup", "":
```
Comment (boumenot) [inline common/multistep_runner.go]: Why support for ""?  Why would you get into this state?

File: common/multistep_runner.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	case "abort":
+		for i, step := range steps {
+			steps[i] = abortStep{step, ui}
+		}
+	case "ask":
+		for i, step := range steps {
+			steps[i] = askStep{step, ui}
+		}
+	default:
+		ui.Error(fmt.Sprintf("Ignoring unknown on-error value %q", config.PackerOnError))
```
Comment (boumenot) [inline common/multistep_runner.go]: Consider adding the name of the configuration value to make the error message clearer for end-users if they provide a bad value.

File: common/multistep_runner.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+func askPrompt(ui packer.Ui) askResponse {
+	for {
+		line, err := ui.Ask("[C]lean up and exit, [A]bort without cleanup, or [R]etry step (build may fail even if retry succeeds)? [car]")
+		if err != nil {
+			log.Printf("Error asking for input: %s", err)
+		}
+
+		switch {
+		case len(line) == 0 || line[0] == 'c':
```
Comment (boumenot) [inline common/multistep_runner.go]: I would make this a case-insensitive comparison, and remove the redundant "[car]" from the message above.

**Comment by boumenot** [comment]
Azure stuff LGTM.  Thanks for the contribution.

File: common/multistep_runner.go (hunk 2 above)
Comment (orivej) [inline common/multistep_runner.go]: Removed `""` case.

File: common/multistep_runner.go (hunk 3 above)
Comment (orivej) [inline common/multistep_runner.go]: Replaced `on-error` with `-on-error`, if this is what you meant.

File: common/multistep_runner.go (hunk 4 above)
Comment (orivej) [inline common/multistep_runner.go]: Done.

File: common/multistep_runner.go (hunk 3 above)
Comment (orivej) [inline common/multistep_runner.go]: Now replaced it with `"Ignoring -on-error=%q argument: unknown on-error value"`

**Comment by rickard-von-essen** [comment]
1) We should fail on unknown values.
```
==> parallels-iso: Ignoring -on-error="ASK" argument: unknown on-error value
```

**Comment by orivej** [comment]
Done. Is the PR still missing some documentation?

File: website/source/docs/command-line/build.html.md [hunk 5]
```diff
@@ -36,7 +36,17 @@ artifacts that are created will be outputted at the end of the build.
...
+    prompt and waits for you to decide to clean up, abort, or retry the failed
+    step.
+
 -   `-only=foo,bar,baz` - Only build the builds with the given
     comma-separated names. Build names by default are the names of their
     builders, unless a specific `name` attribute is specified within
     the configuration.
+
+-   `-parallel=false` - Disable parallelization of multiple builders (on by
+    default).
```
Comment (rickard-von-essen) [inline website/source/docs/command-line/build.html.md]: :+1:

File: website/source/docs/command-line/build.html.md [hunk 6]
```diff
@@ -36,7 +36,17 @@ artifacts that are created will be outputted at the end of the build.
     remove the artifacts from the previous build. This will allow the user to
     repeat a build without having to manually clean these artifacts beforehand.
 
+-   `-on-error=cleanup` (default), `-on-error=abort`, `-on-error=ask` - Selects
+    what to do when the build fails.  `cleanup` cleans up after the previous
+    steps, deleting temporary files and virtual machines.  `abort` exits without
+    any cleanup, but the next build may require `-force`.  `ask` presents a
```
Comment (rickard-von-essen) [inline website/source/docs/command-line/build.html.md]: > `abort` exits without any cleanup, but the next build may require `-force`.

Maybe rephrase to:

`abort` exits without any cleanup, which might require the next build to use `-force`.

File: common/multistep_runner.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+			if _, ok := state.GetOk(multistep.StateCancelled); ok {
+				return askCleanup
+			}
+		}
+	}
+}
+
+func askPrompt(ui packer.Ui) askResponse {
+	for {
+		line, err := ui.Ask("[C]lean up and exit, [a]bort without cleanup, or [r]etry step (build may fail even if retry succeeds)?")
```
Comment (rickard-von-essen) [inline common/multistep_runner.go]: > `[C]lean up and exit, [a]bort without cleanup, or [r]etry step (build may fail even if retry succeeds)?`

I think I ratter liked it to look like:

`[c] Clean up and exit, [a] abort without cleanup, or [r] retry step (build may fail even if retry succeeds)?`

Which makes it a bit more easy to read.

File: command/build.go [hunk 8]
```diff
@@ -284,6 +288,9 @@ Options:
   -except=foo,bar,baz        Build all builds other than these
   -force                     Force a build to continue if artifacts exist, deletes existing artifacts
   -machine-readable          Machine-readable output
+  -on-error=abort            If the build fails, abort without cleanup
+  -on-error=ask              If the build fails, prompt for action
+  -on-error=cleanup          If the build fails, clean up and exit (the default)
```
Comment (rickard-von-essen) [inline command/build.go]: I would prefer if they where on the same line, maybe:

`-on-error=[cleanup|ask|abort]       If the build fails do: clean up (default), ask, or abort`

I think it's clear enough and the details are in the docs.

**Comment by rickard-von-essen** [comment]
> Is the PR still missing some documentation?

No except for the comments I just gave I think this is done (and great).

File: common/multistep_runner.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+	case "abort":
+		for i, step := range steps {
+			steps[i] = abortStep{step, ui}
+		}
+	case "ask":
+		for i, step := range steps {
+			steps[i] = askStep{step, ui}
+		}
+	default:
+		ui.Error(fmt.Sprintf("Ignoring -on-error=%q argument: unknown on-error value", config.PackerOnError))
```
Comment (orivej) [inline common/multistep_runner.go]: I have already deleted the default clause since it can no longer happen. Now `enumflag` makes packer exit with an error like `invalid value "ASK" for flag -on-error: expected one of ["cleanup" "abort" "ask"]`

**Comment by rickard-von-essen** [comment]
LGTM

Thanks for this great improvement!

**Comment by robmorgan** [comment]
thanks @orivej 👍

**Comment by gregorskii** [comment]
Hi,

This was merged into master 25 days ago, and the last release v0.10.2 was cut 16 days ago, but this update is not in the release 0.10.2 on the site, or in homebrew. I had to build from source to get. it. Is this a known issue?

**Comment by rickard-von-essen** [comment]
It's not an issue, it's just read the [CHANGELOG.md](https://github.com/mitchellh/packer/blob/master/CHANGELOG.md#0102-september-20-2016) or:

```bash
 $ git log1 --graph --decorate=full -10 v0.10.2 v0.10.1                                                                                    14s  Sat Oct 15 22:25:18 2016
* 9e8a3fa (tag: refs/tags/v0.10.2) bump gem deps and www version
* fb1c968 Cut version 0.10.2
* 1414401 Cut version 0.10.2
* bb92eef Cut version 0.10.2
* 489c75a Go\'s -X linker flag now requires only one argument (#3540)
* 7bbc29d Fix go vet casing issue
* 8e7bb5e Change version.go to 0.10.1 release
| * 4e5f651 (tag: refs/tags/v0.10.1) Cut version 0.10.1
| * 2c27461 Change version.go to 0.10.1 release
|/
* aa47c90 Udpdated test/race timeouts to 2m because AWS seems to be taking a while
```

v0.10.2 is a rebuild of v0.10.1.

**Comment by gregorskii** [comment]
Ah ok, will pay more attention to the change logs!. Will await 0.10.3. Also this feature works great. 👍

**Comment by orivej** [comment]
@gregorskii Thanks! Which mode do you use: `abort`, `ask`, or both?

**Comment by gregorskii** [comment]
Currently ask, as I am debugging some `googlecompute` Ansible and Chef work, was pulling my hair out last weekend debugging roles as they would fail, and delete the instance before I could SSH in and poke around my work to see if the correct files and config was in place. Having `ask` when building out Packer files is a life saver.

**Comment by lklepner** [comment]
It would be great if -on-error=ask spit out a SSH key and IP addresses along with the prompt. This would make it easy to jump into the build to do some debugging prior to selecting cleanup or abort.

**Comment by petemounce** [comment]
Similarly, for Windows, if the communicator is `winrm` and a password was generated, it would be really handy to dump that password to output so it can be used to log in to the failed build to inspect.

**Comment by SwampDragons** [comment]
Given that the main conversation on this branch happened last year, it's a lot easier for us to track feature requests if you open a new ticket -- or, even better, a PR :)
